wgpu-core/src/instance.rs

```diff
@@ -153,120 +148,42 @@ pub struct DeviceDescriptor {
     pub limits: Limits,
 }
 
-bitflags::bitflags! {
-    #[repr(transparent)]
-    #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
-    pub struct BackendBit: u32 {
-        const VULKAN = 1 << Backend::Vulkan as u32;
-        const GL = 1 << Backend::Gl as u32;
-        const METAL = 1 << Backend::Metal as u32;
-        const DX12 = 1 << Backend::Dx12 as u32;
-        const DX11 = 1 << Backend::Dx11 as u32;
-        /// Vulkan + METAL + DX12
-        const PRIMARY = Self::VULKAN.bits | Self::METAL.bits | Self::DX12.bits;
-        /// OpenGL + DX11
-        const SECONDARY = Self::GL.bits | Self::DX11.bits;
-    }
-}
-
-impl From<Backend> for BackendBit {
-    fn from(backend: Backend) -> Self {
-        BackendBit::from_bits(1 << backend as u32).unwrap()
-    }
-}
-
-pub enum AdapterInputs<'a, I> {
-    IdSet(&'a [I], fn(&I) -> Backend),
-    Mask(BackendBit, fn() -> I),
-}
+impl<F: IdentityFilter<AdapterId>> Global<F> {
+    pub fn enumerate_adapters(&self, inputs: AdapterInputs<F::Input>) -> Vec<AdapterId> {
+        let (adapters, backend_ids) = Adapter::enumerate(self, inputs);
 
-impl<I: Clone> AdapterInputs<'_, I> {
-    fn find(&self, b: Backend) -> Option<I> {
-        match *self {
-            AdapterInputs::IdSet(ids, ref fun) => ids.iter().find(|id| fun(id) == b).cloned(),
-            AdapterInputs::Mask(bits, ref fun) => {
-                if bits.contains(b.into()) {
-                    Some(fun())
-                } else {
-                    None
-                }
-            }
-        }
+        adapters
+            .into_iter()
+            .map(|adapter| adapter.register(self, backend_ids.clone()))
```

Comment:
I'm not entirely sure what `register_identity` does or if there is any alternative, any feedback is welcome.
Also I'm cloning `BackendIds` and I'm not sure what or how big it is.